### PR TITLE
Add EIP-5625: NFT Metadata JSON Schema dStorage Extension

### PR DESCRIPTION
**Pull request by gavfu**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Please fix the CI errors.

**Comment by Pandapip1** [comment]
Side note - why is this needed? Couldn't you just use an `ipfs:` or `arweave:` URL?

**Comment by gavfu** [comment]
> Side note - why is this needed? Couldn't you just use an `ipfs:` or `arweave:` URL?

Good question. The `dStorage` property is not to replace the `ipfs://xxx` or `arweave://xxx` URL, but is targeted to provide complementary information to answer questions like:

- Is this NFT asset stored in a dStorage platform as [Ethereum community suggests](https://ethereum.org/en/developers/docs/storage)?
- If it is, in which way does this dStorage platform guarantees the decentralized storage. For example, what is the persistence mechanism, is there a challenge mechanism to assure data retention, etc.
- Specifically, how to verify this NFT asset is actually stored in that dStorage platform? This is why we need a `dstorage_note`. It could be an Arweave tx, Filecoin deal id, Crust place_storage_order tx, pinning job id of some *IPFS pinning service* like Pinata, etc. 

I also want to elaborate a bit more about IPFS. IPFS is a great protocol to store NFT assets, but sooner or later, more and more people would realize a simple fact: **IPFS does not have built-in incentive mechanism, hence strictly speaking, IPFS storage does not mean Decentralized Storage.** Say, if an NFT asset is stored on some IPFS nodes operated by a single company or organization, the asset will be inaccessible, or even lost, if these IPFS nodes are shutdown. This is also what some IPFS incentive protocols like Crust and Filecoin come for.

In a summary, `dStorage` serves as a complementary to `ipfs://xxx` or `arweave://xxx` URL, since only `ipfs` does not necessarily means `dStorage`, and also we need some `dstorage note` to verify the storage for some true dStorage platforms like `arweave`.

**Comment by Pandapip1** [comment]
> The dStorage property is not to replace the ipfs://xxx or arweave://xxx URL, but is targeted to provide complementary information...

Ah, got it. That does make a bit more sense now.

> IPFS does not have built-in incentive mechanism, hence strictly speaking, IPFS storage does not mean Decentralized Storage. 

First of all, IPFS is most definitely decentralized. It's about as decentralized as the internet (if not more!). The internet doesn't have incentive mechanisms, it's just assumed that people will buy routers. The same applies to IPFS nodes.

Here's a point I think a lot of people miss - while IPFS doesn't have a built-in incentive mechanism, there's an easy way to add one: ensure that the value of the NFT you're minting exceeds the cost of storing its metadata. Therefore, NFT holders are incentivized to store their NFTs' metadata.

Anyways, let's continue this discussion on FEM: https://ethereum-magicians.org/t/eip-5625-nft-metadata-json-schema-dstorage-extension/10754

File: EIPS/eip-5625.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-09-08
+requires: 721, 1155
+---
+
+## Abstract
+
+This standard is an extension to NFT metadata JSON schema defined by [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md). As a complement, it provides additional information about how the NFT asset is stored in a decentralized storage system, if applicable.
```
Comment (Pandapip1) [inline EIPS/eip-5625.md:16]:
```suggestion
This EIP extends the NFT metadata JSON schema defined in [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md), adding a `dStorage` key that provides information about how the NFT data is stored. 
```
I recommend you use this rephrasing of your current abstract. As written, the second sentence is grammatically incorrect.

File: EIPS/eip-5625.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+		"platform": "Crust",
+    "description": "Crust Network, the Incentive Layer of IPFS",
+    "persistence_mechanism": "contract-based",
+    "challenge_mechanism": "mpow",
+    "consensus": "gpos",
+    "dstorage_note": "0x466ab180124de1718d30cd2e018e7fe013a07c4860674110ccd13e97eb31ae16"
+	}
+}
+```
+
```
Comment (Pandapip1) [inline EIPS/eip-5625.md:187]:
```suggestion
```
This should either be in the reference implementation section or omitted. If you would prefer your EIP to be merged more quickly, I would recommend just omitting it.

File: EIPS/eip-5625.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+## Rationale
+
+### Choice between Interface and JSON Schema Extension
+
+A smart contract interface like `NFTdStorageMetadata` would require extra implementation, which is detrimental to the standard adoption, especially for NFT projects that already have their NFT smart contracts finalized and deployed. By contrast, an optional JSON schema extension is noninvasive, and more easily to be adopted.
```
Comment (Pandapip1) [inline EIPS/eip-5625.md:192]:
```suggestion
An extension of the EIP-721 or EIP-1155 contract interfaces would unnecessarily require additional code to implement, and would not be available for use by NFT projects that already have their NFT smart contracts finalized and deployed. An optional JSON schema extension is noninvasive, and more easily adopted.
```
I recommend you rephrase this as above

File: EIPS/eip-5625.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+# Backwards Compatibility
+
+This EIP is backward compatible with [EIP-721](./eip-721.md)  and [EIP-1155](./eip-1155.md) .
```
Comment (Pandapip1) [inline EIPS/eip-5625.md:196]:
```suggestion
This EIP is backward compatible with [EIP-721](./eip-721.md)  and [EIP-1155](./eip-1155.md).
```
Remove trailing space.

File: EIPS/eip-5625.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+## Security Considerations
+
+There are no security considerations related directly to the implementation of this standard.
```
Comment (Pandapip1) [inline EIPS/eip-5625.md:200]:
```suggestion
Needs discussion.
```
I wouldn't be able to say *how* exactly this could cause any security considerations (maybe a malicious contract says that the metadata is immutable, but it's not?), but a draft isn't considered stable enough to definitively say that there *are* ***no*** security considerations.

File: EIPS/eip-5625.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+As a refresher, let's review existing NFT metadata JSON schema standards. [EIP-721](./eip-721.md) defines a standard contract method `tokenURI` to return a given NFT's metadata JSON file, conforming to the *[EIP-721](./eip-721.md) Metadata JSON Schema*, which defines three properties: `name`, `description` and `image`.
+
+Similarly, [EIP-1155](./eip-1155.md) also defines a standard contract method `uri` to return NFT metadata JSON files conforming to the *[EIP-1155](./eip-1155.md) Metadata JSON Schema*, which defines properties like `name`, `decimals`, `description`, `image`, `properties`, `localization`, etc.
+
+Besides, as the world's largest NFT marketplace nowadays, OpenSea defines their own *Metadata Standards*, including a few more properties like `image_data`, `external_url`, `attributes`, `background_color`, `animation_url`, `youtube_url`, etc. This standard is de facto respected and followed by other NFT marketplaces like LooksRare.
+
+Apparently, none of these standards conveys storage information about the mime type asset that NFT token actually represents. This proposal is an effort to fill the missing part.
```
Comment (Pandapip1) [inline EIPS/eip-5625.md:28]: I spot a few grammar issues here. If you run this through something like Grammarly and fix the issues, I won't have to nitpick this and your EIP will be merged much quicker.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Comment by gavfu** [comment]
> Here's a point I think a lot of people miss - while IPFS doesn't have a built-in incentive mechanism, there's an easy way to add one: ensure that the value of the NFT you're minting exceeds the cost of storing its metadata. Therefore, NFT holders are incentivized to store their NFTs' metadata.

Yah, this is a good point we can discuss further on FEM. For example, NFT holders may need to operate some IPFS nodes by themselves, or pay to store the assets in some *IPFS pinning service* like Pinata. What is the checklist of necessary steps to assure data retention on IPFS network? And, after NFT ownership transfer, did new NFT holder need to pay again for the storage, etc.

File: EIPS/eip-5625.md (hunk 6 above)
Comment (gavfu) [reply EIPS/eip-5625.md]: Really appreciate your review. This section is checked via Grammarly and grammar issues are fixed accordingly. I'm not a native English speaker or writer, so if there are any more grammar issues to fix, or some place to rephrase, please don't hesitate to point out.  :)

**Review by Pandapip1 — APPROVED** [review]
The language is a bit colloquial, but currently, I think this is fine for a draft.
